### PR TITLE
Create 'solver' fixture for tests, and update `test_complex_assembler.py` to use it.

### PR DESCRIPTION
**Pull request by chrisrichardson**
Removes the need for PETSc in complex assembler test.

### CODE REVIEW

File: python/test/conftest.py [hunk 1]
```diff
@@ -136,3 +139,64 @@ def tempdir(request):
...
+        rnorm = rnorm_new
+        if (rnorm / rnorm0 < rtol2):
+            return
+        p.array[:nr] = beta * p.array[:nr] + r
+
+    raise RuntimeError(f"Solver exceeded max iterations ({kmax}).")
+
+
+@pytest.fixture(scope="function")
+def solver():
```
Comment (garth-wells) [inline python/test/conftest.py:189]: Name is very generic - it will only work for SPD problems.

Comment (garth-wells) [inline python/test/conftest.py:189]: Name is very generic - parallel case only work for SPD problems.

Comment (chrisrichardson) [reply python/test/conftest.py:189]: Maybe I should always use CG, even in serial? At least it is consistent then... and I can change the name to `cg_solver`?

Comment (garth-wells) [reply python/test/conftest.py:189]: It's probably better to use the same solver in serial and parallel.

I suggest adding [BiCGSTAB](https://en.wikipedia.org/wiki/Biconjugate_gradient_stabilized_method) for non-symmetric and indefinite problems.

Comment (chrisrichardson) [reply python/test/conftest.py:189]: OK, I will change this to `cg_solver()` - maybe we can save BiCGStab for another PR.

File: python/test/conftest.py [hunk 2]
```diff
@@ -136,3 +138,56 @@ def tempdir(request):
...
+       which can work in serial or parallel for testing use.
+       Not suitable for large problems."""
+
+    # Basic Conjugate Gradient solver
+    def _cg(comm, A, b, x):
+        kmax = 500
+        rtol2 = 10 * np.finfo(x.array.dtype).eps
+
+        def _global_dot(comm, v0, v1):
+            return comm.allreduce(np.dot(v0, v1), MPI.SUM)
```
Comment (garth-wells) [inline python/test/conftest.py:155]: Doesn't look like the correct inner product for the complex-valued case.

File: python/test/conftest.py [hunk 3]
```diff
@@ -136,3 +138,56 @@ def tempdir(request):
...
+
+@pytest.fixture(scope="function")
+def cg_solver():
+    """Simple Conjugate Gradient solver for SPD problems,
+       which can work in serial or parallel for testing use.
+       Not suitable for large problems."""
+
+    # Basic Conjugate Gradient solver
+    def _cg(comm, A, b, x):
+        kmax = 500
```
Comment (garth-wells) [inline python/test/conftest.py:151]: `maxit` is more intuitive, and should probably be a parameter with a default value.

File: python/test/conftest.py [hunk 4]
```diff
@@ -136,3 +138,56 @@ def tempdir(request):
...
+        rtol2 = 10 * np.finfo(x.array.dtype).eps
```
Comment (garth-wells) [inline python/test/conftest.py:152]: This is very tight. 

Should probably be a parameter with a default value.

File: python/test/unit/fem/test_complex_assembler.py [hunk 5]
```diff
@@ -6,111 +6,109 @@
...
-    b.ghostUpdate(addv=PETSc.InsertMode.ADD, mode=PETSc.ScatterMode.REVERSE)
+    b.scatter_reverse(la.InsertMode.add)
 
-    # Create solver
-    solver = PETSc.KSP().create(mesh.comm)
-    x = A.createVecRight()
-    solver.setOperators(A)
-    solver.solve(b, x)
+    u = Function(V, dtype=complex_dtype)
+    cg_solver(MPI.COMM_WORLD, A, b, u.x)
```
Comment (garth-wells) [inline python/test/unit/fem/test_complex_assembler.py:105]: comm should com from `V` or the mesh.

**Review by garth-wells — CHANGES_REQUESTED** [review]

File: python/test/conftest.py (hunk 2 above)
Comment (chrisrichardson) [reply python/test/conftest.py:155]: replacing with `vdot`

File: python/test/conftest.py (hunk 4 above)
Comment (chrisrichardson) [reply python/test/conftest.py:152]: Note this is rtol^2, so about 1e-8 for double. I will make it into a parameter

**Review by mscroggs — APPROVED** [review]

**Review by jorgensd — APPROVED** [review]

**Pull request merged**